OpenRA.Game/ExternalMods.cs

```diff
@@ -107,41 +111,87 @@ internal void Register(Manifest mod)
-				LoadMod(yaml.First().Value);
+			if (registration.HasFlag(ModRegistration.User))
+				sources.Add(Platform.SupportDir);
 
-				Directory.CreateDirectory(supportPath);
-				File.WriteAllLines(Path.Combine(supportPath, key + ".yaml"), yaml.ToLines(false).ToArray());
-			}
-			catch (Exception e)
+			// Make sure the mod is available for this session, even if saving it fails
+			LoadMod(yaml.First().Value);
+
+			foreach (var source in sources.Distinct())
 			{
-				Log.Write("debug", "Failed to register currrent mod metadata");
-				Log.Write("debug", e.ToString());
+				if (!Directory.Exists(source))
+					continue;
+
+				var metadataPath = Path.Combine(source, "ModMetadata");
+
+				try
+				{
+					Directory.CreateDirectory(metadataPath);
+					File.WriteAllLines(Path.Combine(metadataPath, key + ".yaml"), yaml.ToLines(false).ToArray());
+				}
+				catch (Exception e)
+				{
+					Log.Write("debug", "Failed to register current mod metadata");
+					Log.Write("debug", e.ToString());
+				}
 			}
+		}
+
+		/// <summary>
+		/// Removes invalid mod registrations:
+		/// * LaunchPath no longer exists
+		/// * LaunchPath and mod id matches the active mod, but the version is different
+		/// * Filename doesn't match internal key
+		/// * Fails to parse as a mod registration
+		/// </summary>
+		internal void ClearInvalidRegistrations(ExternalMod activeMod, ModRegistration registration)
+		{
+			var sources = new List<string>();
+			if (registration.HasFlag(ModRegistration.System))
+				sources.Add(Platform.SystemSupportDir);
+
+			if (registration.HasFlag(ModRegistration.User))
+				sources.Add(Platform.SupportDir);
 
-			// Clean up stale mod registrations:
-			//  - LaunchPath no longer exists (uninstalled)
-			//  - LaunchPath and mod match the current mod, but version differs (newer version installed on top)
-			List<string> toRemove = null;
-			foreach (var kv in mods)
+			foreach (var source in sources.Distinct())
 			{
-				var k = kv.Key;
-				var m = kv.Value;
-				if (!File.Exists(m.LaunchPath) || (m.LaunchPath == launchPath && m.Id == mod.Id && k != key))
+				var metadataPath = Path.Combine(source, "ModMetadata");
+				if (!Directory.Exists(metadataPath))
+					continue;
+
+				foreach (var path in Directory.GetFiles(metadataPath, "*.yaml"))
 				{
-					Log.Write("debug", "Removing stale mod metadata entry '{0}'", k);
-					if (toRemove == null)
-						toRemove = new List<string>();
+					string modKey = null;
+					try
+					{
+						var yaml = MiniYaml.FromStream(File.OpenRead(path), path).First().Value;
+						var m = FieldLoader.Load<ExternalMod>(yaml);
+						modKey = ExternalMod.MakeKey(m);
+
+						// Continue to the next entry if this one is valid
+						if (File.Exists(m.LaunchPath) && Path.GetFileNameWithoutExtension(path) == modKey &&
+							!(activeMod != null && m.LaunchPath == activeMod.LaunchPath && m.Id == activeMod.Id && m.Version != activeMod.Version))
```

Comment:
only note: Can this happen? Register call before ClearInvalidRegistrations overwrites LaunchPath, Id and Version.